### PR TITLE
chore: update knip config

### PR DESCRIPTION
**Pull request by RebeccaStevens**
## PR Checklist

- [ ] Addresses an existing open issue fix #0
- [ ] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/ts-api-utils/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

When not in production mode, entry files should include test files.
When in production mode, project files should not include test files.

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
Using this config, knip pass in #33 with flag `--production`. It still fails when not running in production mode.

File: knip.jsonc
Comment (JoshuaKGoldberg) [inline knip.jsonc:3]: I'm confused - I wouldn't think it right to include test files as entry points. Then any unused code that is tested wouldn't be detected.

Can you talk more about how+why this is useful? E.g. should we be running Knip in _two_ modes: production and test?

Comment (RebeccaStevens) [reply knip.jsonc:3]: > any unused code that is tested wouldn't be detected.

When running knip in production mode, test files won't be entry points and thus we'll detect this case then.
When not in production mode, we'd want to detect code not used at all. So production code, but also test helper function such as those in `src/test/utils.ts`.

> should we be running Knip in _two_ modes: production and test?

Yes, I think so.

**Caveat:**

I haven't use knip before, these are just my initial thoughts on how it should work.

Comment (RebeccaStevens) [reply knip.jsonc:3]: Without `"src/**/*.test.ts"` as an entry point and with `"src/**/*.ts"` as the project files, I would of thought that `createNode` would be flagged as not used, as it's only imported by test files. This don't seem to be the case though, so I'm not entirely sure how knip determines not used code.

Comment (JoshuaKGoldberg) [reply knip.jsonc:3]: Ah cool thanks that makes sense! CCing @webpro in case you have time to weigh in here, since you've already been helpful on this repo 😄 (no worries if not!)

Comment (RebeccaStevens) [reply knip.jsonc:3]: We might want to also exclude `src/test` from production project files:

```json
"project": [
	"src/!(test)/**/!(*.test).ts!",
	"src/!(*.test).ts!",
	"src/**/*.ts"
]
```

**Comment by JoshuaKGoldberg** [comment]
Marking this as blocked on resolving discussion in https://github.com/webpro/knip/issues/62 - in case it ends up impacting us here.

**Comment by webpro** [comment]
Just saying, in the meantime you could add the specific files as `entry` files to unblock this

**Comment by RebeccaStevens** [comment]
> in the meantime you could add the specific files as `entry` files

File: .github/workflows/knip.yml
Comment (JoshuaKGoldberg) [inline .github/workflows/knip.yml:16]: I've been putting one command per workflow `.yml` when possible, so that we can get a granular display of which things are failing. But also these two commands are pretty related.

Personally I'd lean towards splitting these into two files, but I wouldn't block on it. What do you prefer?

File: knip.jsonc
Comment (JoshuaKGoldberg) [inline knip.jsonc:8]: 🤔 I can't get `lint:knip:production` to fail on anything. I:
* Created a standalone file imported by a used file with an unused export
* Created a standalone file imported by a test with an unused export
* Created an unused file with an unused export
...nothing. How does `lint:knip:production` fail?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Requesting changes only in that I'd like to see how to make `lint:knip:production` fail, the code LGTM!

**Comment by RebeccaStevens** [comment]
> I'd like to see how to make `lint:knip:production` fail

If you make a new file (`src/foo.ts`) and put an export in it (`export const foo = 1`), both the version of `knip` will report this unused exported. If you then create a test file for that file (`stc/foo.test.ts`) and import the export (`import { foo } from "./foo"`), now only the production version of `knip` will fail.

At least that's how it should work. I did some testing though and it seems like something with the glob patterns isn't work as expected. I'll look into this.

File: knip.production.jsonc
Comment (RebeccaStevens) [inline knip.production.jsonc:5]: @webpro It doesn't seem to currently be possible to use a single config file for our use case.

Essentially, we want this config, but with the comment addressed.

```jsonc
{
	"$schema": "https://unpkg.com/knip@next/schema.json",
	"entry": ["src/**/index.ts!", "src/**/*.test.ts"],
	"project": [
		"src/**/*.ts!",
		"!src/test/**/*.ts!", // <-- We want these "not" globs to only apply in production mode.
		"!src/**/*.test.ts!"  // <-- Using a separate production specific config seems to be the only way to currently achieve this.
	]
}
```

When using config that runs logic (e.g. `knip.ts`), it would be great if we could access a `isProduction` variable.

File: knip.jsonc
Comment (RebeccaStevens) [reply knip.jsonc]: https://github.com/JoshuaKGoldberg/ts-api-utils/pull/40#issuecomment-1429030453

Should work now with latest commit.

File: .github/workflows/knip.yml
Comment (RebeccaStevens) [reply .github/workflows/knip.yml:16]: I don't mind either way. I'll split these workflows up.

**Review by JoshuaKGoldberg — APPROVED** [review]
💯 thanks!

Whenever you're happy to merge, let's do it!

File: knip.production.jsonc
Comment (webpro) [reply knip.production.jsonc:5]: Sorry for the late reply, but somehow today this popped up in the back of my mind. Any chance you could give it a shot with the latest Knip v1? I think you don't need **any** configuration anymore for either dev or production mode :)

**Pull request merged**
